include/mbgl/map/map_observer.hpp

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+namespace mbgl {
+
+class MapObserver {
+public:
+    static MapObserver& nullObserver() {
+        static MapObserver mapObserver;
+        return mapObserver;
+    }
+
+    virtual void onRegionWillChange() {}
+    virtual void onRegionWillChangeAnimated() {}
+    virtual void onRegionIsChanging() {}
+    virtual void onRegionDidChange() {}
+    virtual void onRegionDidChangeAnimated() {}
+    virtual void onWillStartLoadingMap() {}
+    virtual void onDidFinishLoadingMap() {}
+    virtual void onDidFailLoadingMap() {}
+    virtual void onWillStartRenderingFrame() {}
+    virtual void onDidFinishRenderingFrame() {}
+    virtual void onDidFinishRenderingFrameFullyRendered() {}
+    virtual void onWillStartRenderingMap() {}
```

Comment:
`onWillStartRenderingMap()` and `onDidFinishRenderingMap()` are based on MapKit’s [`-[MKMapViewDelegate mapViewWillStartRenderingMap:]`](https://developer.apple.com/reference/mapkit/mkmapviewdelegate/1451970-mapviewwillstartrenderingmap?language=objc) (“when one or more tiles are revealed and require rendering”) and [`-[MKMapViewDelegate mapViewDidFinishRenderingMap:fullyRendered:]`](https://developer.apple.com/reference/mapkit/mkmapviewdelegate/1451897-mapviewdidfinishrenderingmap?language=objc) (“finishes rendering all of the currently visible tiles to the best of its ability”), which have analogues in our iOS and macOS SDKs.

If our iOS documentation is to be believed, `on*RenderingFrame()` would be called on each frame of a camera animation or property transition, even where no additional tiles need to be fetched and displayed. However, I don’t know for sure if that’s how we’ve implemented these notifications, given #2775.

/cc @friedbunny